utils.sh

```diff
@@ -2,6 +2,22 @@
 
 set -o pipefail
 
+function retry_with_timeout() {
+  retries=$1
+  timeout_duration=$2
+  command=${@:3}
```

Comment:
$@ is an array, but here you are assigning to a string variable, because later on you use it as a string parameter to bash. Better use $*

```suggestion
  command=${*:3}
```